ether/vivaldi.py

```diff
@@ -0,0 +1,81 @@
+import random
+from typing import Tuple
+
+import numpy as np
+
+from ether.core import Node, Coordinate
+
+c_e = 0.9
+"a tuning parameter that influences the weight of the current error in each cycle"
+c_c = 0.25
+"tuning parameter that modulates the force"
+dimensions = 8
+"dimensionality of the vector space"
+
+max_error = 1.5
+min_height = 10e-6
+
+
+class VivaldiCoordinate(Coordinate):
+    """
+    Coordinate holds network coordinates and local error, used by the Vivaldi coordinate mapping algorithm.
+    """
+    position: np.ndarray
+    height: float
+    error: float
+    vivaldi_runs: int
+
+    def __init__(self, position: np.ndarray = None, height: float = None, error: float = None):
+        super().__init__()
+        self.position = position if position is not None else np.array([0.0] * dimensions)
+        self.height = height if height is not None else min_height
+        self.error = error or max_error
+        self.vivaldi_runs = 0
+
+    def __repr__(self) -> str:
+        return f'Coordinate({self.position}, height={self.height}, error={self.error})'
+
+    def apply_force(self, force: float, other: 'VivaldiCoordinate'):
+        unit, norm = _unit_vector_at(self.position, other.position)
+        self.position += unit * force
+        if norm > 0:
+            self.height += (self.height + other.height) * force / norm
+            self.height = max(self.height, 10e-3)
+
+    def distance_to(self, other: 'VivaldiCoordinate'):
+        return np.linalg.norm(self.position - other.position) + self.height + other.height
+
+
+def _unit_vector_at(v1: np.ndarray, v2: np.ndarray) -> Tuple[np.ndarray, float]:
+    """
+    returns a unit vector pointing at v1 from v2
+    """
+    result = v1 - v2
+    norm = np.linalg.norm(result)
```

Comment:
couldn't this be moved into the first `if` block?, so between line 55 and 56?

Reply:
sure, but I wasn't sure whether making it a static method or putting it in global scope was the better choice